src/payment/data/sagas.js

```diff
@@ -19,10 +19,38 @@ import { handleErrors, handleMessages } from '../../feedback';
 export function* handleFetchBasket() {
   try {
     yield put(fetchBasket.request());
-    const result = yield call(PaymentApiService.getBasket);
+    let result = yield call(PaymentApiService.getBasket);
     yield put(fetchBasket.success(result));
     yield put(basketDataReceived(result));
     yield call(handleMessages, result.messages, true);
+    if (result.products.length === 1) {
+      const [product] = result.products;
+      const { courseKey } = product;
+      const discount = yield call(
+        PaymentApiService.checkBasketForDiscount,
+        courseKey,
+      );
+      if (discount.discount_applicable) {
+        try {
+          yield put(fetchBasket.request());
+          result = yield call(
+            PaymentApiService.getBasket,
+            discount.jwt,
+          );
+          yield put(fetchBasket.success(result));
```

Comment:
when we need to update the page after the discount is applied @abutterworth just suggested one option of just copying the code to be run a second time
but if we need to pull something out to a separate function to be more DRY that's fine too